Adding disable property caching FAQ

Taha-Firoz opened the pull request:
The property caching feature wasn't well documented. This commit adds it to the FAQ section of the book.

zeenix commented on book/src/faq.md:
```diff
@@ -1,5 +1,18 @@
 # FAQ
 
+## Why aren't property values updating for my service that doesn't notify changes?
+
+A common issue might arise when using zbus and your service's property values aren't updating. 
+This is due to zbus' default caching policy, which updates the value of a property only when 
```
This is specific to `Proxy` but the text makes it sound like it's the general behavior (e.g you won't face this with the low-level `MessageStream` API). Would be good to be specific here.

zeenix commented on book/src/faq.md:
```diff
@@ -1,5 +1,18 @@
+By default, if your service does not emit change notifications, the property values will not 
+update accordingly. 
+
+However, you can explicitly bypass this caching mechanism by using the 
+`#[dbus_proxy(property(emits_changed_signal = "false"))]` annotation. 
```
Similarly here, we're assuming that it's a specific proxy type generated by `dbus_proxy`. One can have the same issue with `Proxy`. We should mention that [`ProxyBuilder::uncached_properties` method](https://docs.rs/zbus/3.13.1/zbus/struct.ProxyBuilder.html#method.uncached_properties) can be used when dealing with `Proxy`.

zeenix commented on book/src/faq.md:
```diff
@@ -1,5 +1,18 @@
+However, you can explicitly bypass this caching mechanism by using the 
+`#[dbus_proxy(property(emits_changed_signal = "false"))]` annotation. 
+This way, zbus will not expect change notifications for the property, allowing for a 
+manual update of the values.
```
Would be good to add here:

> Please refer to `emits_changed_signal` part of [`dbus_proxy` documentation](https://docs.rs/zbus/3.13.1/zbus/attr.dbus_proxy.html) for other values of this attribute and their purpose.

zeenix commented on book/src/faq.md:
```diff
@@ -1,5 +1,18 @@
 # FAQ
 
+## Why aren't property values updating for my service that doesn't notify changes?
```
Please add this new question at the bottom. :)

zeenix reviewed (changes requested):
Thanks, LGTM but can be improved. While you fix your commit, please take this opportunity to add an emoji too. :laughing:

zeenix commented on book/src/faq.md:
```diff
@@ -101,6 +88,27 @@ involve converting the methods in question to take `&self` argument instead of `
+By default, if your service does not emit change notifications, the property values will not
+update accordingly. 
+
+However, you can explicitly bypass this caching mechanism for a client:
+### 1. Client built via `dbus_proxy`
+By adding the `#[dbus_proxy(property(emits_changed_signal = "false"))]` annotation to your proxy
+
+### 2. Client built via `ProxyBuilder`
```
* Some vertical spacing please.
* bypassing -> disabling caching for specific properties.
* A few other rewordings (see below).

```suggestion
However, you can disabling caching for specific properties:

### 1. `dbus_proxy`-generated proxies

By adding the `#[dbus_proxy(property(emits_changed_signal = "false"))]` annotation to your property.

### 2. `Proxy` instances

```

zeenix replied:
Oh and I think it's best to also list disabling of property caching for all the properties of a proxy through `ProxyBuilder::cache_properites` (which works for both kinds of proxies.

zeenix commented on book/src/faq.md:
```diff
@@ -101,6 +88,27 @@ involve converting the methods in question to take `&self` argument instead of `
+By adding the `#[dbus_proxy(property(emits_changed_signal = "false"))]` annotation to your proxy
+
+### 2. Client built via `ProxyBuilder`
+`ProxyBuilder::uncached_properties` method to your builder
```
this phrase doesn't make sense. Please write a full sentence and end with a `.`.

zeenix commented on book/src/faq.md:
```diff
@@ -101,6 +88,27 @@ involve converting the methods in question to take `&self` argument instead of `
+This way, zbus will not expect change notifications for the property, allowing for a 
+manual update of the values.
+
+For more information about all the possible values for `emits_changed_signal` refer to [`dbus_proxy`](https://docs.rs/zbus/3.13.1/zbus/attr.dbus_proxy.html) documentation.
```
break the line at `to` to fit under 100 chars.

zeenix commented on book/src/faq.md:
```diff
@@ -101,6 +88,27 @@ involve converting the methods in question to take `&self` argument instead of `
+
+For more information about all the possible values for `emits_changed_signal` refer to [`dbus_proxy`](https://docs.rs/zbus/3.13.1/zbus/attr.dbus_proxy.html) documentation.
+
+
```
Redundant line.

```suggestion
```

zeenix reviewed (changes requested):
Getting close. :) Apart from those nitpicks, please squash fixup commits to the unmerged commit you're fixing, [as per the contribution guide](https://github.com/dbus2/zbus/blob/main/CONTRIBUTING.md#submitting-pull-requests).

P.S. How do I convince you to add emojis (other than providing a simple/easy tool [that I already did](https://toot.cat/@zeenix/110420611919155218)? ;))?

zeenix commented:
@Taha-Firoz will you be finishing this soon?

Taha-Firoz commented:
Oh yup, sorry I was caught up with my job. I'll try getting to it this week

Taha-Firoz commented:
> ... please squash fixup commits to the unmerged commit you're fixing...

I'm very sorry I tried very hard to do this but I just failed I think, I'm only used to committing and pushing and this rebasing stuff was very hard to wrap my head around. I did an interactive rebase, I squashed the most recent 2 commits and picked the first one and added a new commit message. I really don't understand why there are 5 commits now 😭😭😭😭

zeenix commented:
> I'm only used to committing and pushing and this rebasing stuff was very hard to wrap my head around.

That's most people on Github so you're not alone. However, basic git it an important skill that will serve you well in life so I strongly suggest not giving up.  Most Linux-specific projects (systemd, linux kernel, libvirt, etc etc) will ask for the same so it's good to have these skill under your belt.

> I did an interactive rebase, I squashed the most recent 2 commits and picked the first one and added a new commit message. I really don't understand why there are 5 commits now

Did you check the log/history of your branch before pushing? If you did a forced push, the history should be identical. Try rebasing again.

Fortunately this is one logical change so I can "squash and merge" if/when you want to give up. I just want to give you the opportunity to learn rebasing for any future contributions.

Taha-Firoz commented:
I am very sorry for adding these 15 commits 😭😭😭😭 I am still trying to figure it out.

Taha-Firoz commented:
Okay I think I got it, I didn't do a force push after I had done the rebase which lead to the additional commits. The 15 commits... well I don't know what happened then I was trying to undo what I had done and somehow that happened. 

Finally I did a hard reset to the rebase commit and did a force push.

zeenix commented:
> I am very sorry for adding these 15 commits sobsobsobsob I am still trying to figure it out.

It's fine to make mistakes. That's how you learn. :)



> Finally I did a hard reset to the rebase commit and did a force push.

:+1:

zeenix commented on book/src/faq.md:
```diff
@@ -88,6 +88,37 @@ involve converting the methods in question to take `&self` argument instead of `
+
+However, you can disabling caching for specific properties:
+
+### 1. `dbus_proxy` generated proxies
+
+Add the `#[dbus_proxy(property(emits_changed_signal = "false"))]` annotation to the property
+for which you desire to disable caching on.
+
+### 2. `Proxy` instances
+
+Use the `ProxyBuilder::uncached_properties` method on your Client built with `ProxyBuilder` to 
+disable caching on it's properties.
+
+### 3. Either type
```
Not sure we need headings here.

zeenix commented on book/src/faq.md:
```diff
@@ -88,6 +88,37 @@ involve converting the methods in question to take `&self` argument instead of `
+In order to disable caching for either type of Client use the `ProxyBuilder::cache_properites` 
+method.
+
+This way, zbus will not expect change notifications for the property, allowing for a 
+manual update of the values.
```
I'd just remove this. It's obvious from above.

zeenix reviewed (changes requested):
OK, now that you've commits sorted and you know how to fix commits, let's fix the final few things before we merge this.

Taha-Firoz commented on book/src/faq.md:
```diff
@@ -88,6 +88,37 @@ involve converting the methods in question to take `&self` argument instead of `
+
+### 2. `Proxy` instances
+
+Use the `ProxyBuilder::uncached_properties` method on your Client built with `ProxyBuilder` to 
```
Client is just the terminology I've seen used more when referring to an object that's interfacing with a remote service. However you're correct that proxy is more accurate here.

zeenix reviewed (approved):
Still see many small issues actually but I'm afraid if I point them out, you'll loose your motivation to continue working on this.

Thanks for the contribution.

Taha-Firoz commented:
@zeenix No no no, I love this entire process. I really appreciate your effort and I'm learning a lot. I'll be faster now

zeenix commented:
> @zeenix No no no, I love this entire process. I really appreciate your effort and I'm learning a lot. I'll be faster now

Great to hear!

Taha-Firoz commented:
So are you reopening it?

zeenix commented:
> So are you reopening it?

Re-opening what? It's already merged. :)

zeenix commented on book/src/faq.md:
```diff
@@ -88,6 +88,28 @@ involve converting the methods in question to take `&self` argument instead of `
+- Add the `#[dbus_proxy(property(emits_changed_signal = "false"))]` annotation to the property
+for which you desire to disable caching on.
+
+- Use `ProxyBuilder` to build your proxy instance and use `ProxyBuilder::uncached_properties` method
+to list all properties you wish to disable caching for.
+
+- In order to disable caching for either type of proxy use the `ProxyBuilder::cache_properites` 
+method.
```
`ProxyBuilder::uncached_properties` and `ProxyBuilder::cache_properites` should be links to the docs on these methods.

Taha-Firoz replied:
Yup I followed the new convention of adding doc URLS

zeenix commented on book/src/faq.md:
```diff
@@ -88,6 +88,28 @@ involve converting the methods in question to take `&self` argument instead of `
+method.
+
+For more information about all the possible values for `emits_changed_signal` refer
+ to [`dbus_proxy`](https://docs.rs/zbus/3.13.1/zbus/attr.dbus_proxy.html) documentation.
```
* Best to use footnote link syntax. See other links in this file (e.g `tctiog` and `Type`).
* Just realized that this para only applies to the first bullet point. Please move this sentence there and remove this para.

Taha-Firoz replied:
I added a block quote to it because just moving it up to the bullet point looked weird, I am not sure if putting it in a block quote was appropriate 😅

zeenix replied:
> I added a block quote to it because just moving it up to the bullet point looked weird,

What block quote? :thinking: I am just talking about the whole paragraph (don't see any quotation involved). It only applies to one bullet point so it belongs there.

Taha-Firoz replied:
So I just append this to the first bullet point?

zeenix replied:
yes, the sentence, as I wrote in my first comment here. :)

zeenix commented on book/src/faq.md:
```diff
@@ -88,6 +88,28 @@ involve converting the methods in question to take `&self` argument instead of `
+
 [tctiog]: https://github.com/tokio-rs/tokio/issues/2201
 [`Type`]: https://docs.rs/zvariant/3.1.0/zvariant/derive.Type.html
 [`SerializeDict`]: https://docs.rs/zvariant/3.0.0/zvariant/derive.SerializeDict.html
```
your branch was a bit old so you didn't see my changes from a637c7b87 so it's understandable you linked to exact versions on docs.rs but you can keep in mind when updating/adding links now.

zeenix reviewed (commented):
@Taha-Firoz Doing a post-merge review. Please create a new PR to address these points. Remember to keep your commits atomic and now that your changes from this PR are merged already, you're not going to be fixing commits anymore (only on the new PR if I find anything to nitpick about :)).
